substrate/frame/meta-tx/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Meta Tx
+//!
+//! TODO docs
+
+#![cfg_attr(not(feature = "std"), not_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub use pallet::*;
+
+use frame_support::{
+	dispatch::{DispatchInfo, GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+	traits::OriginTrait,
+};
+use frame_system::pallet_prelude::*;
+use sp_runtime::traits::TransactionExtension;
+use sp_runtime::traits::TransactionExtensionBase;
+use sp_runtime::traits::{Dispatchable, IdentifyAccount, Verify};
+
+/// Meta transaction type.
+// TODO: The `MetaTx` type is similar to `sp_runtime::generic::UncheckedExtrinsic`. However,
+// `MetaTx` cannot replace generic::UncheckedExtrinsic because we need to box the call type,
+// given that `MetaTx` is used as an argument type for a call.
+#[derive(Encode, Decode, PartialEq, Eq, TypeInfo, Clone, RuntimeDebug)]
+pub struct MetaTx<Address, Signature, Call, TxExtension> {
+	proof: Proof<Address, Signature>,
+	call: Box<Call>,
+	tx_ext: TxExtension,
+}
+
+impl<Address, Signature, Call, TxExtension> MetaTx<Address, Signature, Call, TxExtension> {
+	pub fn new_signed(
+		address: Address,
+		signature: Signature,
+		call: Call,
+		tx_ext: TxExtension,
+	) -> Self {
+		Self { proof: Proof::Signed(address, signature), call: Box::new(call), tx_ext }
+	}
+}
+
+#[derive(Encode, Decode, PartialEq, Eq, TypeInfo, Clone, RuntimeDebug)]
+pub enum Proof<Address, Signature> {
+	Signed(Address, Signature),
+	// TODO `General` as in `sp_runtime::generic::Preamble`.
+}
```

Comment:
I think we can branch off the `UncheckedExtrinsic` and `Preamble` logic a bit here and consider introducing a variant for any type of custom validation logic supported by the pallet. In `TransactionExtension`s, they would be individual extensions in the pipeline, one for each validation type, each wrapping a `General` transaction.

Here, we care only about the call and validation type, which is nice.